Agent merging

maxxxxxdlp opened the pull request:
Fixes [#1030](https://github.com/specify/specify7/issues/1030)
Fixes https://github.com/specify/specify7/issues/2692

maxxxxxdlp commented:
Pretty much finished. Now just to polish it

<img width="1800" alt="Screenshot 2022-12-26 at 19 35 27" src="https://user-images.githubusercontent.com/40512816/209597522-1052d847-6329-4450-bab4-93ac0df51e71.png">


<img width="1800" alt="Screenshot 2022-12-26 at 19 35 35" src="https://user-images.githubusercontent.com/40512816/209597520-00457792-8b53-4bfe-8bb8-3ca96b6c4b68.png">

grantfitzsimmons commented:
After merging three agents, I get a 404 error above the query builder.

<img width="1273" alt="Screenshot 2022-12-27 at 6 37 45 AM" src="https://user-images.githubusercontent.com/37256050/209667775-a014b63d-7bd8-47a0-83d3-f7f1672f6557.png">

https://newmexico-agent-merging.test.specifysystems.org/specify/query/14/

grantfitzsimmons commented:
![image](https://user-images.githubusercontent.com/37256050/209668152-61ecbda2-88d8-4081-b298-f6bf37d033cf.png)

After merging 3 agents, only one of the items disappears and one is left behind (outlined in red). When going to that record, it does not exist.

grantfitzsimmons commented:
https://user-images.githubusercontent.com/37256050/209668742-f77fd0ba-1215-4805-ba74-a2b2d5909313.mp4

Example workflow when merging a large batch of agents.

Frequent errors in the console reporting 

```
{type: 'invalidResponseCode', statusText: Array(3), responseText: 'No Agent matches the given query.'}
```

Takes quite a while to complete the merging, and shows a dialog for a second that is confusing to the user after beginning the merge.

> This DB is really good for testing this on (https://newmexico-agent-merging.test.specifysystems.org/specify/query/14/)

grantfitzsimmons commented:
After selecting 3 agents in a new tab (after closing all other tabs) and after all actions appeared to be finished, I get an out of date error.

<img width="1279" alt="image" src="https://user-images.githubusercontent.com/37256050/209669208-f828284e-be62-44aa-9abc-28dbb3b11296.png">

```
Agent object 47591 is out of date
```

[Specify 7 Crash Report - 2022-12-27T12_50_44.370Z.txt](https://github.com/specify/specify7/files/10308229/Specify.7.Crash.Report.-.2022-12-27T12_50_44.370Z.txt)

---

Can be recreated by running the query and selecting the following 3 results:
<img width="1254" alt="image" src="https://user-images.githubusercontent.com/37256050/209669371-57596bc3-c41a-438e-bc4d-08455785546b.png">

Recieved this error upon recreation:
```
Agent object 47590 is out of date
```

grantfitzsimmons commented:
The issue where a new merge dialog appears while merging is taking place with the 404 error after the action is taken is something that needs to be resolved

grantfitzsimmons commented:
https://user-images.githubusercontent.com/37256050/209698864-031b8038-ee79-40d1-9074-957471febea2.mp4

Same second dialog problem happening after initiating the merge

grantfitzsimmons reviewed (changes requested):
Same second dialog problem happening after initiating the merge

maxxxxxdlp commented:
> Same second dialog problem happening after initiating the merge

This is because an error has occurred (you see that you had 4 columns, but the final dialog has only 3 - one record was merged successfully, the next one failed)

The problem is that it doesn't show you the error messages - that was a bug introduced on this branch and fixed. The bug appears fixed for me (tried on coldfish on test server). Maybe the server just didn't update on time when you tested. Could you try this again?

maxxxxxdlp commented:
I forgot to mention during the demo that agent merging is also available though the Form Meta menu.

Reasons:
1. Allows people with no query builder access to use it
2. It's going to be a well-advertised feature of Specify 7 - we need to include it in more than one place in the UI

chanulee1 commented:
What needs to be tested on this branch?

maxxxxxdlp commented:
Besides the logic for properly handling business rules, agent merging is now feature complete.

Context:
Agent merging allows to merge duplicate agent merging
In the future, we would enable this feature for tables other than Agent
You can access this feature by selecting multiple agents in the query results or by opening an agent form, opening Form Meta menu, and clicking on "Merge Record"

To test:
Interface for merging records
Interface for merging -to-many relationships (merging Agent Addresses or Agent members)
Pressing the "Merge" button would do merging record by record. If it fails, a dialog will be shown. The dialog should be dismissible. Once you resolve the business rule violation mentioned in the error dialog, the records should be mergable.
After merging, the duplicates should be removed from the query results. Similarly, if you opened "Record Merging" from the form, after merging all, it should open the final record in a form. If you were in a record set while doing that, the final record should also be opened in a context of a record set.

Also, I once had this error happen:
<img width="1386" alt="Screenshot 2023-01-13 at 09 27 32" src="https://user-images.githubusercontent.com/40512816/213542192-5d6e7ee8-c59b-4f66-8228-cb80625bd4cb.png">

Please tell if you also have it happen.

maxxxxxdlp commented:
Agent Merging dialog is now more polished (based on the feedback I received). Changes:
1. The references to this record section is cleaned up (now it skipps showing tables like Loan Agent and goes directly to Loan, and instead of displaying each resource separately, displays a count - clicking on it shows a dialog like in "Browse in Forms")
2. Timestamps are shown as relative dates
3. Added vertical line to the right of the final record to make it more distinct (also made the heading bold and renamed the button labels in the "Preview" row)
4. Auto merge no longer runs when you open the dialog. Instead, you can run it on demand by pressing a button
5. There is an "x" button in the heading of each resource that allows to dismiss that resource from merging (if you found out that it's not actually a duplicate)

<img width="1434" alt="Screenshot 2023-01-19 at 21 58 50" src="https://user-images.githubusercontent.com/40512816/213614545-90fb31ac-bf7b-414f-8f87-6c3c20a02682.png">

chanulee1 commented:
Errors when attempting to merge certain agents with any other agent record, such as `User, Test C` and `Bentley, Andrew C`. Errors occur while the 'References to this record' section is loading.

`User, Test C`:
![image](https://user-images.githubusercontent.com/103299204/213785762-644092b5-f72b-4444-ab33-55803f004221.png)

[Specify 7 Crash Report - 2023-01-20T18_54_19.275Z.txt](https://github.com/specify/specify7/files/10469374/Specify.7.Crash.Report.-.2023-01-20T18_54_19.275Z.txt)

`Bentley, Andrew C`:
![image](https://user-images.githubusercontent.com/103299204/213786102-9e702a1e-f18f-44bd-af47-4c8e1e1ba2dd.png)

[Specify 7 Crash Report - 2023-01-20T19_14_41.904Z.txt](https://github.com/specify/specify7/files/10469378/Specify.7.Crash.Report.-.2023-01-20T19_14_41.904Z.txt)

chanulee1 commented:
'References to this record' section overflows for some agent records.
![image](https://user-images.githubusercontent.com/103299204/213786897-e47cc658-ded0-48eb-abda-d90cfb187122.png)

chanulee1 commented:
Is this expected behavior? Deleting one of the address records in the merged agent record also deletes the corresponding address record for a duplicate agent when the address merging interface is closed. Unable to get this record back without restarting the agent merging interface.

https://user-images.githubusercontent.com/103299204/213796938-73e343fe-fe4d-4029-915c-1be6ef20d68c.mp4

chanulee1 commented:
Deleting multiple records in merged record indents field labels in -to-many relationship merging interface.

https://user-images.githubusercontent.com/103299204/213800927-321396f0-95e5-49a3-9ad8-a4d0bf68a34e.mp4

chanulee1 commented:
Merging two agent records with respective address records creates agent record with three address records (Two copies of records from Duplicate 2). Also, when no value is chosen for the 'first name' field of the merged record, the query results display the Duplicate 1's value for this field after merging. Going to the merged record shows that this value was not saved to this field.

https://user-images.githubusercontent.com/103299204/213814146-f2dc6a77-c218-41d8-8837-f68c71ddf23a.mp4

Similar behavior when merging more than two agents.

https://user-images.githubusercontent.com/103299204/213820223-6099c5ed-602c-4bca-82e3-9983b84f64e2.mp4

maxxxxxdlp commented:
> Errors when attempting to merge certain agents with any other agent record...

Fixed

> 'References to this record' section overflows for some agent records.

Fixed

Fixed [#2736](https://github.com/specify/specify7/issues/2736)
Fixed [#2874](https://github.com/specify/specify7/issues/2874)
Fixed [#2876](https://github.com/specify/specify7/issues/2876)
Fixed [#2877](https://github.com/specify/specify7/issues/2877)

Also, the delete blockers dialog on the forms has been modified to exactly match the "Usages" section in agent merging - need testing. If you open the delete blockers dialog, then edit any of the related resources so that they no longer point back at the current resource (i.e, reassign Determiner in Determination), that Determination should be removed from the delete blockers dialog.

Will fix the other bugs in the next few days

maxxxxxdlp commented:
> Deleting multiple records in merged record indents field labels in -to-many relationship merging interface.

Fixed in 57a95397c
Fixed https://github.com/specify/specify7/issues/2879

maxxxxxdlp commented:
> Is this expected behavior? Deleting one of the address records in the merged agent record also deletes the corresponding address record for a duplicate agent when the address merging interface is closed. Unable to get this record back without restarting the agent merging interface.

Not expected behavior. Fixed

> Deleting multiple records in merged record indents field labels in -to-many relationship merging interface.

Fixed

> the query results display the Duplicate 1's value for this field after merging

Grant reported that as https://github.com/specify/specify7/issues/2887

> Going to the merged record shows that this value was not saved to this field.

Fixed in https://github.com/specify/specify7/issues/2879

> Merging two agent records with respective address records creates agent record with three address records (Two copies of records from Duplicate 2)

Good catch! The cause is a bit related to https://github.com/specify/specify7/issues/2713, so I will append your comment there

maxxxxxdlp commented:
Subsequent agent merging bugs will be tracked in this project as this pull request is getting to large - https://github.com/specify/specify7/projects/15#card-87804911

(issues in the `To Do` column are aimed for first release, issues in the `Ready on branch` column are fixed and ready for testing)

maxxxxxdlp commented on specifyweb/frontend/js_src/lib/components/Merging/index.tsx:
```diff
@@ -251,6 +257,51 @@ function Merging({
+      mergeResource,
+      'blockersChanged',
+      (): void => {
+        const onlyDeferredBlockers = Array.from(
```
this logic seems like something that would be commonly used. can you turn it into a helper function on the SaveBlocker class?

maxxxxxdlp replied:
on second thought, how is this different from `SaveBlocker.hasOnlyDeferredBlockers` function?

maxxxxxdlp commented on specifyweb/frontend/js_src/lib/components/Merging/index.tsx:
```diff
@@ -251,6 +257,51 @@ function Merging({
+  const [saveBlocked, setSaveBlocked] = React.useState(false);
+  const [showSaveBlockedDialog, setShowBlockedDialog] = React.useState(false);
+
+  React.useEffect(() => {
```
We already have a hook `useSaveBlockers`. Can you check if that works for you here?
If not, can you modify that hook to work for you?

Also check the other place where you added a blockers change listener hook to see if you can replace it with `useSaveBlockers`

maxxxxxdlp commented on specifyweb/frontend/js_src/lib/components/Merging/index.tsx:
```diff
@@ -251,6 +257,51 @@ function Merging({
+
+  return (
+    <>
+      {!saveBlocked ? (
```
don't negate `if` conditions. instead swap the branches

maxxxxxdlp commented on specifyweb/frontend/js_src/lib/components/Merging/index.tsx:
```diff
@@ -251,6 +257,51 @@ function Merging({
+      {!saveBlocked ? (
+        <Submit.Blue form={formId}>{treeText.merge()}</Submit.Blue>
+      ) : (
+        <Submit.Red className="cursor-not-allowed">
```
while you make the button look not clickable, does it actually prevent the button click from submitting the form?

also, try disabling the button to see if you like the styling of the disabled button (might remove the need for adding `cursor-not-allowed` manually)

melton-jason replied:
> while you make the button look not clickable, does it actually prevent the button click from submitting the form?
Yes, this always prevents the button click from submitting the form. I don't pass in the formId into the form prop for this purpose. 

When the formId is passed in as a form prop, there _is_ a bug where the user can submit the form where there are errors (saveblockers or validation errors) in any subview. 
I now know to take a look at the browser API and ensure the parent resource is invalid when a Merging Subview is invalid

maxxxxxdlp commented on specifyweb/frontend/js_src/lib/components/Merging/index.tsx:
```diff
@@ -251,6 +257,51 @@ function Merging({
+          {treeText.merge()}
+        </Submit.Red>
+      )}
+      {showSaveBlockedDialog && (
```
looks like a bug. you call `setShowBlockedDialog(false)`, but you never call `setShowBlockedDialog(true)` thus this dialog will never be displayed

maxxxxxdlp reviewed (commented):
Looks like GitHub doesn't allow me to request changes on a pull request because I'm the author of the pull request, so I am leaving these as comments instead.

Looking good and a lot of great progress. Next thing would be to look into handling the back-end business rule error response and attaching that error to a field

don't hesitate if you have questions/need design help

maxxxxxdlp commented on specifyweb/frontend/js_src/lib/components/DataModel/__tests__/businessRules.test.ts:
do you want to close that pull request in favor of this one, or do you want to resolve those comments under that pull request first?

melton-jason replied:
> do you want to close that pull request in favor of this one, or do you want to resolve those comments under that pull request first?

I have been thinking about this. It would be nice keeping #2965 open, as developments during testing can be kept separated. 
However, this would also require cherry-picking commits over to the `refactor-busrules`  branch when business rule changes are made on this branch, which is not ideal. 

Ultimately I think the benefits for having everything separate for testing procedures (and to keep both Pull Requests much cleaner) outweigh the extra work needed to ensure that Business Rules are synchronized between the branches. 

To reduce workload, I think it is wise to 'finish' Business Rules- or at the very least, everything that Agent Merging needs for Business Rules)- on this branch so all of the needed commits can be cherry-picked in one setting to the `refactor-busrules` branch.

melton-jason replied:
I've cherry-picked all of the Business Rule related commits to the `refactor-busrule` branch. 
The business rule files `businessRules.ts`, `businessRuleDefs.ts`, and `uniqueness_rules.json` are now mostly synchronized, There was one commit which _also_ modified Agent Merging related files, 2eac871, which I had to exclude from the process.

melton-jason commented on specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts:
```diff
@@ -60,7 +60,7 @@ type MappedBusinessRuleDefs = {
   BorrowMaterial: {
     fieldChecks: {
-      quantityReturned: (
+      quantityreturned: (
```
I have toyed around with using camelCase instead of lowercase for the fieldCheck definitions. 

The only issue is that we would have to convert the fieldnames coming from BackBone-which are in lowercase- _back_ to camelCase which is substantially more difficult than the other way around. 

To invoke the correct rule, we need to be able to use the BackBone field names (in lowecase) to pull the right field from fieldCheck definitions. Here is where the issue occurs in the code:

https://github.com/specify/specify7/pull/2965/files#diff-afde7250d8079dd81ae9657ed120cf46a1f5147a88ac908f3d20840d79d88ee2R369

maxxxxxdlp replied:
The .getField() and .getFields() methods on SpecifyModel class are perfect for converting lowercase names back to camel case

Also, keep in mind that we are planning to get rid of backbone (and it's lowercase fields too), thus don't contain your design by the limitations of a legacy soon to be removed system

melton-jason replied:
Great! `.getField()` sounds exactly like what I need then. Thank you!

melton-jason replied:
Fixed in 373d4ba

melton-jason commented on specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts:
```diff
@@ -349,16 +349,22 @@ export const nonUniqueBusinessRuleDefs: MappedBusinessRuleDefs = {
   },
 };
 
+// From this code, Typescript believes that a businessRuleDefs uniqueIn can be from any table
```
I've made sure the comments are consistent in businessRules.ts and businessRuleDefs.ts in 373d4ba

melton-jason commented on specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts:
```diff
@@ -28,7 +28,7 @@ export class BusinessRuleManager<SCHEMA extends AnySchema> {
 
   private addPromise(
-    promise: Promise<BusinessRuleResult | string | undefined>
+    promise: Promise<BusinessRuleResult | string | void | undefined>
```
I have removed all unnecessary `void` statements in 373d4ba

melton-jason commented on specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts:
```diff
@@ -125,26 +125,26 @@ export class BusinessRuleManager<SCHEMA extends AnySchema> {
-      | RA<{ field: string; otherFields: string[] }> =
+    const scopeFields =
       this.rules?.uniqueIn !== undefined
         ? this.rules?.uniqueIn[fieldName as Lowercase<TableFields<SCHEMA>>] ??
```
Fixed in https://github.com/specify/specify7/commit/373d4bae0ab1de2d8f3e6c98a46bf7f01359430d

melton-jason commented on specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts:
```diff
@@ -98,8 +98,8 @@ export class BusinessRuleManager<SCHEMA extends AnySchema> {
+  ): Promise<RA<void | null>> {
     return Promise.all(
       results.map((result) => {
         if (!result) return null;
```
I have removed (or replaced with undefined) all unnecessary nulls in https://github.com/specify/specify7/commit/373d4bae0ab1de2d8f3e6c98a46bf7f01359430d.
The only remaining place that uses null that is not in an if-statement is in the type for the `uniqueness_rules.json` file. 

When a value should be unique at a global (institution) level, then we say it should be unique to `null`. 

The nulls are entirely self-contained however. When the uniqueness rules are being assigned to the final `businessRuleDefs` object, I replace these nulls with undefined.

melton-jason commented on specifyweb/frontend/js_src/lib/components/DataModel/businessRules.ts:
```diff
@@ -354,23 +352,30 @@ export class BusinessRuleManager<SCHEMA extends AnySchema> {
-    if (this.rules === undefined) {
+  ): Promise<BusinessRuleResult | undefined | void> {
+    if (this.rules === undefined || ruleName === 'uniqueIn') {
       return Promise.resolve(undefined);
```
Fixed in https://github.com/specify/specify7/commit/373d4bae0ab1de2d8f3e6c98a46bf7f01359430d
